ARCHITECTURE.md

```diff
@@ -0,0 +1,162 @@
+
+In integration test, we have a sample JSON file. We generate another JSON file from by calling Main.main with args. And compare it using `git diff --no-index`. This is the same strategy we use in `jnigen` tests to compare generated bindings.
+
+### Formatting
+All Java files should be formatted with [google-java-format](https://github.com/google/google-java-format/), including the examples committed in jnigen tests.
+
+`google-java-format` can be installed using `scoop` on windows and `homebrew` on MacOS. On Linux, you can download the JAR and write a shell script like this:
+
+```sh
+#!/bin/sh
+java -jar ~/.local/path_to_jar/google-java-format.jar $@
+```
+There's a plugin available for IntelliJ.
+
+The command line JAR apparently needs a manual listing of java files, and does not process directories recursively. You can use something like `$(find jnigen/java/ -name *.java)` as an argument when running from shell.
+
+See [google-java-format#129](https://github.com/google/google-java-format/issues/129).
+
+## `package:jnigen`
+This is the component which generates the bindings. It invokes summarizer, ([summary.dart](jnigen/lib/src/summary/summary.dart)), and processes each class to generate bindings.
+
+### Preprocessing
+Before being passed to one of the `BindingsGenerator` classes, class declaration structs are _preprocessed_. Preprocessingassigns a shorter name (which can also be final visible name in `single_file` config). Shorter name is the simple name of the class + optionally a number to disambiguate naming conflicts. This leads to shorter bindings than using FQN, while still being pretty deterministic (due to sorting in the summarizer). Additionally, in practice, Java libraries do not have many colliding names, because Java classes are also imported unqualified.
+
+Preprocessing also renames methods (since dart doesn't support overloading). Ad-hoc overloading can lead to a methods signature being different than the same-named method in its superclass - which is an error. We actually do some circus to prevent this. We save java signature of each method as a string, which is a unique representation. Now it's possible to have a map per class of these methods and their associated disambiguating numbers. By ensuring the superclass is sorted before subclass, we ensure methods with same signature get the same number. (Sounds like topological sort, but actually it's not needed. We just lazily preprocess the parent and mark it as preprocessed.)
+
+### File mapping
+Depending on `output` >> `dart` >> `structure` key in config, output can be package structured or single file. In package structure, the original source is mirrored, with each top level class being mapped to a separate dart file. An `_init.dart` file will be created to hold common variables. In each package, `_package.dart` file will be created, which exports all classes directly (__not transitively__) in the package. This is equivalent of the wildcard import in Java.
```

Comment:
config -> the config
output -> the output
single file -> a single file
package structure -> the package structure
equivalent of -> equivalent to